codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt

```diff
@@ -605,13 +611,164 @@ private class ServerHttpProtocolImplGenerator(
+            }
+    }
+
+    private fun queryParamsTargetMapValueType(targetMapValue: Shape): QueryParamsTargetMapValueType =
+        if (targetMapValue.isStringShape) {
+            QueryParamsTargetMapValueType.STRING
+        } else if (targetMapValue.isListShape) {
+            QueryParamsTargetMapValueType.LIST
+        } else if (targetMapValue.isSetShape) {
+            QueryParamsTargetMapValueType.SET
+        } else {
+            throw ExpectationNotMetException("""
+                @httpQueryParams trait applied to non-supported target
+                $targetMapValue of type ${targetMapValue.type}
+                """.trimIndent(),
+                targetMapValue.sourceLocation)
+        }
+
+    private fun serverRenderQueryStringParser(writer: RustWriter, operationShape: OperationShape) {
+        val queryBindings =
+            httpBindingResolver.requestBindings(operationShape).filter {
+                it.location == HttpLocation.QUERY
+            }
+        // Only a single structure member can be bound to `httpQueryParams`, hence `find`.
+        val queryParamsBinding =
+            httpBindingResolver.requestBindings(operationShape).find {
+                it.location == HttpLocation.QUERY_PARAMS
+            }
+        if (queryBindings.isEmpty() && queryParamsBinding == null) {
+            return
+        }
+
+        fun HttpBindingDescriptor.queryParamsBindingTargetMapValueType(): QueryParamsTargetMapValueType  {
+            check(this.location == HttpLocation.QUERY_PARAMS)
+            val queryParamsTarget = model.expectShape(this.member.target)
+            val mapTarget = queryParamsTarget.asMapShape().get()
+            return queryParamsTargetMapValueType(model.expectShape(mapTarget.value.target))
+        }
+
+        with(writer) {
+            rustTemplate("""
+                let query_string = request.uri().query().ok_or(#{SmithyHttpServer}::rejection::MissingQueryString)?;
+                let pairs = #{SerdeUrlEncoded}::from_str::<Vec<(&str, &str)>>(query_string)?;
+                """.trimIndent(),
+                *codegenScope
+            )
+
+            if (queryParamsBinding != null) {
+                rustTemplate("let mut query_params: #{HashMap}<String, " +
+                        "${queryParamsBinding.queryParamsBindingTargetMapValueType().asRustType().render()}> = #{HashMap}::new();",
+                    "HashMap" to RustType.HashMap.RuntimeType,
+                )
+            }
+            val (queryBindingsTargettingCollection, queryBindingsTargettingSimple) =
+                queryBindings.partition { model.expectShape(it.member.target) is CollectionShape }
+            queryBindingsTargettingSimple.forEach {
+                rust("let mut seen_${it.memberName.toSnakeCase()} = false;")
+            }
+            queryBindingsTargettingCollection.forEach {
+                rust("let mut ${it.memberName.toSnakeCase()} = Vec::new();")
```

Comment:
Note here that we always use a `Vec`. If the member shape targets a Smithy set, we're handing `Vec`s to the user anyway. See the discussion in #962.

This means however that we're not deduplicating in case the query string contains repeated values with the same key, and the user chose a Smithy set.

Reply:
I filed https://github.com/awslabs/smithy-rs/issues/972